sdcflows/transform.py

```diff
@@ -72,151 +183,189 @@ def fit(self, spatialimage):
 
         """
         # Calculate the physical coordinates of target grid
-        if isinstance(spatialimage, (str, bytes, Path)):
-            spatialimage = nb.load(spatialimage)
-
-        # Initialize xform (or set identity)
-        xfm = self.xfm if self.xfm is not None else nt.linear.Affine()
+        if isinstance(target_reference, (str, bytes, Path)):
+            target_reference = nb.load(target_reference)
+
+        approx = (
+            approx if affine is not None else False
+        )  # Approximate iff affine is defined
+        affine = affine if affine is not None else np.eye(4)
+        target_affine = target_reference.affine.copy()
+
+        # Project the reference's grid onto the fieldmap's
+        target_reference = target_reference.__class__(
+            target_reference.dataobj,
+            affine @ target_affine,
```

Comment:
When you call `fit(in_data, fmap2data_xfm, approx)`, that gets mapped to `(target_reference, affine, approx)`.

`target_affine` maps target IJK onto target RAS. `affine` maps fmap RAS onto target RAS. The only sensible thing to do here is to map fmap RAS onto target IJK, so we need:

```suggestion
            np.linalg.inv(affine) @ target_affine,
```

Alternately, `fmap2data_xfm` needs to be inverted before passing.

Reply:
I believe this is correct - I agree the nomenclature and the docstring are not clear.

`affine` is the result of calling a registration tool such as ANTs configured for a linear transform with at most 12 degrees of freedom and called with the image carrying `target_affine` as reference and the fieldmap reference as moving.

The result of such a registration framework is an affine (our `affine` here) that maps coordinates in reference (target) RAS on to fieldmap RAS. I'm quite confident this inversion is not needed, but I agree we want to implement some tests here. As I mentioned to @mgxd, I have a good dataset for that :)

For now, what I can do is improving the docstring with something along the lines I have described here. The docstrings need all be revised anyway.

Reply:
Sounds good. We need very clear docstrings then. My worry with the second fit called during apply is that this step is being undone in any case, so getting it wrong or right will not have an discernible impact until that bug is fixed.

Reply:
I have added one extra input to the interface, xor'ed with the original one, to pass the reverse transform if that's what you have. Then, inside the interface it is inverted.